------------

# Комментарии

Привет, Сергей!

Я проверила твою работу и оставила в ней комментарии. Комментарии оформлены в таких ячейках и вставлены сразу после ячейки или группы ячеек, к которым относятся. А в самом низу общие комментарии по работе. Надеюсь, тебе все будет понятно :)

Ты справился с проектом, поздравляю)

------------

Открытие файла
-----
Открываем файл __.csv__ с данными по адрессу _/datasets/data.csv_ 
Вывели общую информацию: название столбцев, количество, тип данных.
Вывели первые 5 строк на экран для понимания, как представленны данные

In [1]:
import pandas as pd
data_loan_users = pd.read_csv('/datasets/data.csv')
print(data_loan_users.info())
print()
data_loan_users.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None



children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

------------

# Комментарии

Отлично, данные загружены, есть их краткое описание)

Мой совет: выводи первые строки таблицы в отдельной ячейке и без использования функции __print__ - это позволит увидеть таблицу в нормальном виде, без переноса столбцов на следующии строки. Также, строки таблицы и информацию о данных в таблице лучше выводить в разных ячейках - это облегчит восприятие информации)

------------

В столбце 'education' большое количество дубликатов. Убираем их с помощью функции str.lower() 

In [2]:
data_loan_users['education'] = data_loan_users['education'].str.lower()
data_loan_users.groupby('education').count()

children  days_employed  dob_years  education_id  \
education                                                               
высшее                   5260           4716       5260          5260   
начальное                 282            261        282           282   
неоконченное высшее       744            675        744           744   
среднее                 15233          13693      15233         15233   
ученая степень              6              6          6             6   

                     family_status  family_status_id  gender  income_type  \
education                                                                   
высшее                        5260              5260    5260         5260   
начальное                      282               282     282          282   
неоконченное высшее            744               744     744          744   
среднее                      15233             15233   15233        15233   
ученая степень                   6                 6       6            6   

                      debt  total_income  purpose  
education                                          
высшее                5260          4716     5260  
начальное              282           261      282  
неоконченное высшее    744           675      744  
среднее              15233         13693    15233  
ученая степень           6             6        6

Как видно, дубликаты убраны и данные из столбика __'education'__ можно использовать для анализа

------------

# Комментарии

Отлично) избавились от дубликатов категорий в разных регистрах)

Не хватает описания того, почему именно такой метод избавления от дубликатов выбран (указания типа дубликатов).

Здесь для вывода можно было не использовать функцию __print__ . Достаточно просто написать выражение в скобках и получить таблицу более приятную для прочтения) 

------------

Проверяе таблицу на дубликаты

In [3]:
data_loan_users.duplicated().sum()


71

In [4]:
data_loan_users[data_loan_users.duplicated()]

children  days_employed  dob_years education  education_id  \
2849          0            NaN         41   среднее             1   
3290          0            NaN         58   среднее             1   
4182          1            NaN         34    высшее             0   
4851          0            NaN         60   среднее             1   
5557          0            NaN         58   среднее             1   
...         ...            ...        ...       ...           ...   
20702         0            NaN         64   среднее             1   
21032         0            NaN         60   среднее             1   
21132         0            NaN         47   среднее             1   
21281         1            NaN         30    высшее             0   
21415         0            NaN         54   среднее             1   

          family_status  family_status_id gender income_type  debt  \
2849    женат / замужем                 0      F   сотрудник     0   
3290   гражданский брак                 1      F   пенсионер     0   
4182   гражданский брак                 1      F   сотрудник     0   
4851   гражданский брак                 1      F   пенсионер     0   
5557   гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
20702   женат / замужем                 0      F   пенсионер     0   
21032   женат / замужем                 0      F   пенсионер     0   
21132   женат / замужем                 0      F   сотрудник     0   
21281   женат / замужем                 0      F   сотрудник     0   
21415   женат / замужем                 0      F   пенсионер     0   

       total_income                            purpose  
2849            NaN            покупка жилья для семьи  
3290            NaN                    сыграть свадьбу  
4182            NaN                            свадьба  
4851            NaN                            свадьба  
5557            NaN                    сыграть свадьбу  
...             ...                                ...  
20702           NaN         дополнительное образование  
21032           NaN              заняться образованием  
21132           NaN                       ремонт жилью  
21281           NaN  покупка коммерческой недвижимости  
21415           NaN                  операции с жильем  

[71 rows x 12 columns]

------------

# Комментарии

Такой формат вывода (без __print__) кажется более удобным для восприятия, как думаешь?

------------

Удаляем дубликаты с помощью функции .drop_duplicates() и проверяем на наличие дубликатов

In [5]:
data_loan_users = data_loan_users.drop_duplicates().reset_index(drop=True)                                 

In [6]:
data_loan_users.duplicated().sum()

0

------------

# Комментарии

Отлично) только я не вижу проверки на "наличие дубликатов" :(

------------

Проверяем на отритцательные значения данные

In [7]:
minus_children = data_loan_users[data_loan_users['children'] < 0]['children'].count()
print(minus_children)

47


In [8]:
minus_days_employed = data_loan_users[data_loan_users['days_employed'] < 0]['days_employed'].count()
print(minus_days_employed)

15906


In [9]:
minus_days_dob_years = data_loan_users[data_loan_users['dob_years'] < 0]['dob_years'].count()
print(minus_days_dob_years)

0


In [10]:
minus_days_total_income = data_loan_users[data_loan_users['total_income'] < 0]['total_income'].count()
print(minus_days_total_income)

0


Как видно отритцательны значения есть в столбцах 'children' и 'days_employed'. Просмотрим данные в столбце 'children'

In [11]:
data_loan_users.loc[data_loan_users['children'] < 0]

children  days_employed  dob_years            education  education_id  \
291          -1   -4417.703588         46              среднее             1   
705          -1    -902.084528         50              среднее             1   
742          -1   -3174.456205         57              среднее             1   
800          -1  349987.852217         54              среднее             1   
941          -1            NaN         57              среднее             1   
1363         -1   -1195.264956         55              среднее             1   
1929         -1   -1461.303336         38              среднее             1   
2073         -1   -2539.761232         42              среднее             1   
3812         -1   -3045.290443         26              среднее             1   
4198         -1    -901.101738         41              среднее             1   
4399         -1  398001.302888         64              среднее             1   
4539         -1   -1811.899756         32              среднее             1   
5268         -1   -1143.485347         46              среднее             1   
6008         -1   -1361.258696         46               высшее             0   
6379         -1  370215.476226         48              среднее             1   
7184         -1   -5928.202068         34              среднее             1   
7272         -1    -526.318451         51              среднее             1   
7609         -1            NaN         35              среднее             1   
7679         -1   -3237.360455         53              среднее             1   
8220         -1   -1803.441074         54               высшее             0   
8242         -1   -3113.998449         31              среднее             1   
8552         -1    -195.479496         31              среднее             1   
9553         -1   -2896.629686         34               высшее             0   
9562         -1   -2710.419901         28               высшее             0   
10162        -1   -1743.604011         37              среднее             1   
10350        -1  345774.125957         63               высшее             0   
10900        -1  340499.039342         50              среднее             1   
11090        -1   -3438.463024         59              среднее             1   
11249        -1   -1048.782203         30               высшее             0   
11598        -1    -370.827130         27              среднее             1   
12284        -1   -8493.101252         61  неоконченное высшее             2   
13760        -1            NaN         42              среднее             1   
14328        -1    -268.337037         23              среднее             1   
15114        -1   -9851.184337         44              среднее             1   
15400        -1   -3614.220232         40               высшее             0   
16093        -1    -457.861760         33              среднее             1   
16227        -1   -2802.218127         40               высшее             0   
17054        -1   -2809.693200         34              среднее             1   
17386        -1    -895.379738         37              среднее             1   
17614        -1   -4571.957475         41              среднее             1   
18174        -1   -3575.215641         33              среднее             1   
19048        -1    -617.246968         38              среднее             1   
19311        -1    -372.034749         43              среднее             1   
19360        -1  350340.760224         28              среднее             1   
20328        -1  355157.107212         69              среднее             1   
20650        -1    -661.822321         32               высшее             0   
21071        -1   -1422.668059         44              среднее             1   

               family_status  family_status_id gender  income_type  debt  \
291         гражданский брак                 1      F    сотрудник     0   
705          ж

Убираем пропущенные данные из столбика days_employed. Для этого воспользуемся методом .median(). 
Для данного типа значений лучше использовать медианное значение данных. Так как медиана по стажу 
дает меньшую статистичекскую ошибку

Исходя из данных это скорее всего ошибка при формировании DF исправим ее путем замены отритцательного значения на положительное

In [12]:
data_loan_users.loc[data_loan_users['children'] < 0, 'children'] = data_loan_users.loc[data_loan_users['children'] < 0, 'children']*(-1)

Проверка на отритцательные значения

In [13]:
minus_children = data_loan_users[data_loan_users['children'] < 0]['children'].count()
print(minus_children)

0


In [14]:
data_loan_users['days_employed'] = pd.to_numeric(data_loan_users['days_employed'], errors='coerce')
data_loan_users['days_employed'] = data_loan_users['days_employed'].fillna(data_loan_users['days_employed'].median())
data_loan_users['days_employed'] = data_loan_users['days_employed'].astype(int)
print(data_loan_users['days_employed'].head(10))

0     -8437
1     -4024
2     -5623
3     -4124
4    340266
5      -926
6     -2879
7      -152
8     -6929
9     -2188
Name: days_employed, dtype: int64


Как видим в столбце присутвсуют отритцательные значения и излишневыские значения. для уточнения данных по сталбцу необходимо связаться с сотавителем данных и уточнить что имелось ввиду.

------------

# Комментарии

Хорошо) разобрались с отрицательными значениями)

Да, как вариант можно просто изменить знак у отрицательных значений признака _children_. Хотя, могло получится и так, что отрицательное значение тоже было опцией при заполнении. Также, можно удалить эти отрицательные значения, так как их достаточно мало. 

Как ты понял, что медиана по стажу даст наименьшую статистическую ошибку? Я не вижу, чтобы ты смотрел на разброс значений в данных.

Еще ты проводишь изменение типа данных признака _days__employed_, но никак это не комментируешь - почему?

Хорошая идея - узнать причину отрицательных значений в столбце _days__employed_. Ведь там есть и положительные значений. Интересно, а какой примерно стаж у людей в годах? Может и тут есть аномалии. 

Мне кажется, здесь немного нарушен порядок вывода ячеек с описанием :( Пожалуйста, проверяй последовательность ячеек перед отправкой.

------------

Убираем пропущенные данные из столбика total_income. Для этого воспользуемся методом .median(). Так как в таблице присутсвуют данные столбца income_type, которые в значительной степени коррелируют со значением столбца total_income.Необходимо расчитать медианное значение по значений столбца  total_income в разрезе income_type как это сделано:

In [15]:
median_income_type_un = data_loan_users[data_loan_users['income_type'] == 'безработный']['total_income'].median()
median_income_type_dec = data_loan_users[data_loan_users['income_type'] == 'в декрете']['total_income'].median()
median_income_type_gov = data_loan_users[data_loan_users['income_type'] == 'госслужащий']['total_income'].median()
median_income_type_com = data_loan_users[data_loan_users['income_type'] == 'компаньон']['total_income'].median()
median_income_type_ret = data_loan_users[data_loan_users['income_type'] == 'пенсионер']['total_income'].median()
median_income_type_bus = data_loan_users[data_loan_users['income_type'] == 'предприниматель']['total_income'].median()
median_income_type_emp = data_loan_users[data_loan_users['income_type'] == 'сотрудник']['total_income'].median()
median_income_type_stud = data_loan_users[data_loan_users['income_type'] == 'студент']['total_income'].median()

Сейчас заполняем пропуски

In [16]:
data_loan_users['total_income'] = pd.to_numeric(data_loan_users['total_income'], errors='coerce')
data_loan_users['total_income'] = data_loan_users['total_income'].fillna(0)
data_loan_users.loc[(data_loan_users['total_income'] == 0) & (data_loan_users['income_type'] == 'безработный'), 'total_income'] = median_income_type_un
data_loan_users.loc[(data_loan_users['total_income'] == 0) & (data_loan_users['income_type'] == 'в декрете'), 'total_income'] = median_income_type_dec
data_loan_users.loc[(data_loan_users['total_income'] == 0) & (data_loan_users['income_type'] == 'госслужащий'), 'total_income'] = median_income_type_gov
data_loan_users.loc[(data_loan_users['total_income'] == 0) & (data_loan_users['income_type'] == 'компаньон'), 'total_income'] = median_income_type_com
data_loan_users.loc[(data_loan_users['total_income'] == 0) & (data_loan_users['income_type'] == 'пенсионер'), 'total_income'] = median_income_type_ret
data_loan_users.loc[(data_loan_users['total_income'] == 0) & (data_loan_users['income_type'] == 'предприниматель'), 'total_income'] = median_income_type_bus
data_loan_users.loc[(data_loan_users['total_income'] == 0) & (data_loan_users['income_type'] == 'сотрудник'), 'total_income'] = median_income_type_emp
data_loan_users.loc[(data_loan_users['total_income'] == 0) & (data_loan_users['income_type'] == 'студент'), 'total_income'] = median_income_type_stud
data_loan_users['total_income'] = data_loan_users['total_income'].astype(int)
print(data_loan_users['total_income'].head(10))


0    253875
1    112080
2    145885
3    267628
4    158616
5    255763
6    240525
7    135823
8     95856
9    144425
Name: total_income, dtype: int64


Проверка откорректированных данных

In [17]:
print(data_loan_users.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 12 columns):
children            21454 non-null int64
days_employed       21454 non-null int64
dob_years           21454 non-null int64
education           21454 non-null object
education_id        21454 non-null int64
family_status       21454 non-null object
family_status_id    21454 non-null int64
gender              21454 non-null object
income_type         21454 non-null object
debt                21454 non-null int64
total_income        21454 non-null int64
purpose             21454 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB
None


Как видим мы изменили пропуски в таблице

------------

# Комментарии

Хорошо) пропущенные значения заменены)

Хотелось бы увидеть объяснение выбору метода заполнения пропусков. А также комментарий к тому, что ты заменяешь тип данных в столбце _total__income_.

Не думал, почему появились пропуски?

------------

Анализируем строчные данные на повторы

In [18]:
education = data_loan_users.groupby('education')['debt'].sum()
print(education)

education
высшее                  278
начальное                31
неоконченное высшее      68
среднее                1364
ученая степень            0
Name: debt, dtype: int64


In [19]:
family_status = data_loan_users.groupby('family_status')['debt'].sum()
print(family_status)

family_status
Не женат / не замужем    274
в разводе                 85
вдовец / вдова            63
гражданский брак         388
женат / замужем          931
Name: debt, dtype: int64


In [20]:
family_status = data_loan_users.groupby('family_status')['debt'].sum()
print(family_status)

family_status
Не женат / не замужем    274
в разводе                 85
вдовец / вдова            63
гражданский брак         388
женат / замужем          931
Name: debt, dtype: int64


In [21]:
gender = data_loan_users.groupby('gender')['debt'].sum()
print(gender)

gender
F      994
M      747
XNA      0
Name: debt, dtype: int64


В столбце 'gender' есть пропуск XNA. Проверем данные которые содержатся в строках (строке со значением с значением XNA в столбике  'gender'.

In [22]:
number_str_XNA = data_loan_users.index[data_loan_users.gender == 'XNA']
print(data_loan_users.loc[number_str_XNA])

       children  days_employed  dob_years            education  education_id  \
10684         0          -2358         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10684  гражданский брак                 1    XNA   компаньон     0   

       total_income               purpose  
10684        203905  покупка недвижимости  


Как видно остальные данные заполненны. Т.к. в дальнешем анализ не подразумевает анализ влияние "пола" на наличие долга,
то данную строку оставляем в DF

------------

# Комментарии

Отлично) как я понимаю, повторов нет) 

Хотелось бы увидеть побольше комментариев.

Не думал о причине возникновения значения "XNA"?

------------

Проведем автоматическую лемматизацию

In [23]:
from pymystem3 import Mystem
m = Mystem()
big_str = ''
for row in data_loan_users['purpose']:
    big_str += row
    big_str += ' '
lemmas = m.lemmatize(big_str)
from collections import Counter
print(Counter(lemmas))


Counter({' ': 55023, 'недвижимость': 6351, 'покупка': 5897, 'жилье': 4460, 'автомобиль': 4306, 'образование': 4013, 'с': 2918, 'операция': 2604, 'свадьба': 2324, 'свой': 2230, 'на': 2222, 'строительство': 1878, 'высокий': 1374, 'получение': 1314, 'коммерческий': 1311, 'для': 1289, 'жилой': 1230, 'сделка': 941, 'дополнительный': 906, 'заниматься': 904, 'подержать': 853, 'проведение': 768, 'сыграть': 765, 'сдача': 651, 'семья': 638, 'собственный': 635, 'со': 627, 'ремонт': 607, 'приобретение': 461, 'профильный': 436, 'подержанный': 111, ' \n': 1})


Как видно наиболее часто встречаемые слова: операция 15 117, недвижимость - 5772, жилье - 5039, автомобиль - 4306, получение - 4013, образование' - 4013, 'проведение' - 2324, - 'свадьба': 2324.
Исходя из анализа наилучшими данными для столбца 'purpose':
операции с автомобилем
операции с жильем
операции с недвижемостью
получение образования
проведение свадьбы

In [24]:
purpose = data_loan_users.groupby('purpose')['debt'].count()
print(purpose)

purpose
автомобили                                478
автомобиль                                494
высшее образование                        452
дополнительное образование                460
жилье                                     646
заняться высшим образованием              496
заняться образованием                     408
на покупку автомобиля                     471
на покупку подержанного автомобиля        478
на покупку своего автомобиля              505
на проведение свадьбы                     768
недвижимость                              633
образование                               447
операции с жильем                         652
операции с коммерческой недвижимостью     650
операции с недвижимостью                  675
операции со своей недвижимостью           627
покупка жилой недвижимости                606
покупка жилья                             646
покупка жилья для сдачи                   651
покупка жилья для семьи                   638
покупка коммерческой недви

Категеризируем данные и столбика  purpose

операции с автомобилем  - автомобиль, автомобили, на покупку автомобиля, на покупку подержанного автомобиля, на покупку своего автомобиля, приобретение автомобиля, свой автомобиль, сделка с автомобилем, сделка с подержанным автомобилем 

(Данную категорию включаем любые операции с автомобилем)       
  
операции с жильем       -  жилье, операции с жильем, покупка жилой недвижимости, покупка жилья, покупка жилья для семьи,  покупка своего жилья, строительство жилой недвижимости, ремонт жилью 

(Данную категорию включаем любые операции с жильем) 

опреации с недвижимостью - недвижимость, операции с коммерческой недвижимостью, операции с недвижимостью, операции со своей недвижимостью, покупка жилья для сдачи, покупка коммерческой недвижимости, покупка недвижимости, строительство недвижимости, строительство собственной недвижимости,     

получение образования  - образование, высшее образование, дополнительное образование, заняться высшим образованием, заняться образованием, получение высшего образования, получение дополнительного образования, получение образования, профильное образование        

проведение свадьбы      -  свадьба, на проведение свадьбы, сыграть свадьбу            

In [25]:
data_loan_users.loc[(data_loan_users['purpose'] == 'автомобиль'), 'purpose'] = 'операции с автомобилем'
data_loan_users.loc[(data_loan_users['purpose'] == 'автомобили'), 'purpose'] = 'операции с автомобилем'
data_loan_users.loc[(data_loan_users['purpose'] == 'на покупку автомобиля'), 'purpose'] = 'операции с автомобилем'
data_loan_users.loc[(data_loan_users['purpose'] == 'на покупку подержанного автомобиля'), 'purpose'] = 'операции с автомобилем'
data_loan_users.loc[(data_loan_users['purpose'] == 'на покупку своего автомобиля'), 'purpose'] = 'операции с автомобилем'
data_loan_users.loc[(data_loan_users['purpose'] == 'приобретение автомобиля'), 'purpose'] = 'операции с автомобилем'
data_loan_users.loc[(data_loan_users['purpose'] == 'свой автомобиль'), 'purpose'] = 'операции с автомобилем'
data_loan_users.loc[(data_loan_users['purpose'] == 'сделка с автомобилем'), 'purpose'] = 'операции с автомобилем'
data_loan_users.loc[(data_loan_users['purpose'] == 'сделка с подержанным автомобилем'), 'purpose'] = 'операции с автомобилем'
data_loan_users.loc[(data_loan_users['purpose'] == 'образование'), 'purpose'] = 'получение образования'
data_loan_users.loc[(data_loan_users['purpose'] == 'высшее образование'), 'purpose'] = 'получение образования'
data_loan_users.loc[(data_loan_users['purpose'] == 'дополнительное образование'), 'purpose'] = 'получение образования'
data_loan_users.loc[(data_loan_users['purpose'] == 'заняться высшим образованием'), 'purpose'] = 'получение образования'
data_loan_users.loc[(data_loan_users['purpose'] == 'заняться образованием'), 'purpose'] = 'получение образования'
data_loan_users.loc[(data_loan_users['purpose'] == 'получение высшего образования'), 'purpose'] = 'получение образования'
data_loan_users.loc[(data_loan_users['purpose'] == 'получение дополнительного образования'), 'purpose'] = 'получение образования'
data_loan_users.loc[(data_loan_users['purpose'] == 'получение образования'), 'purpose'] = 'получение образования'
data_loan_users.loc[(data_loan_users['purpose'] == 'профильное образование'), 'purpose'] = 'получение образования'
data_loan_users.loc[(data_loan_users['purpose'] == 'свадьба'), 'purpose'] = 'проведение свадьбы'
data_loan_users.loc[(data_loan_users['purpose'] == 'на проведение свадьбы'), 'purpose'] = 'проведение свадьбы'
data_loan_users.loc[(data_loan_users['purpose'] == 'сыграть свадьбу'), 'purpose'] = 'проведение свадьбы'
data_loan_users.loc[(data_loan_users['purpose'] == 'жилье'), 'purpose'] = 'операции с жильем'
data_loan_users.loc[(data_loan_users['purpose'] == 'операции с жильем'), 'purpose'] = 'операции с жильем'
data_loan_users.loc[(data_loan_users['purpose'] == 'покупка жилой недвижимости'), 'purpose'] = 'операции с жильем'
data_loan_users.loc[(data_loan_users['purpose'] == 'покупка жилья'), 'purpose'] = 'операции с жильем'
data_loan_users.loc[(data_loan_users['purpose'] == 'покупка жилья для семьи'), 'purpose'] = 'операции с жильем'
data_loan_users.loc[(data_loan_users['purpose'] == 'покупка своего жилья'), 'purpose'] = 'операции с жильем'
data_loan_users.loc[(data_loan_users['purpose'] == 'строительство жилой недвижимости'), 'purpose'] = 'операции с жильем'
data_loan_users.loc[(data_loan_users['purpose'] == 'строительство собственной недвижимости'), 'purpose'] = 'операции с недвижимостью'
data_loan_users.loc[(data_loan_users['purpose'] == 'недвижимость'), 'purpose'] = 'операции с недвижимостью'
data_loan_users.loc[(data_loan_users['purpose'] == 'операции с коммерческой недвижимостью'), 'purpose'] = 'операции с недвижимостью'
data_loan_users.loc[(data_loan_users['purpose'] == 'операции с недвижимостью'), 'purpose'] = 'операции с недвижимостью'
data_loan_users.loc[(data_loan_users['purpose'] == 'операции со своей недвижимостью'), 'purpose'] = 'операции с недвижимостью'
data_loan_users.loc[(data_loan_users['purpose'] == 'покупка жилья для сдачи'), 'purpose'] = 'операции с недвижимостью'
data_loan_users.loc[(data_loan_users['purpose'] == 'покупка коммерческой недвижимости'), 'purpose'] = 'операции с недвижимостью'
data_loan_users.loc[(data_loan_users['purpose'] == 'покупка недвижимости'), 'purpose'] = 'операции с недвижимостью'
data_loan_users.loc[(data_loan_users['purpose'] == 'строительство недвижимости'), 'purpose'] = 'операции с недвижимостью'
data_loan_users.loc[(data_loan_users['purpose'] == 'ремонт жилью'), 'purpose'] = 'операции с жильем'

In [26]:
purpose = data_loan_users.groupby('purpose')['debt'].count()
print(purpose)

purpose
операции с автомобилем      4306
операции с жильем           5039
операции с недвижимостью    5772
получение образования       4013
проведение свадьбы          2324
Name: debt, dtype: int64


------------

# Комментарии

Здорово) автоматическая лемматизация сильно упрощает процесс выделения наиболее частых слов)

------------

Создаем таблицы children куда входят столбцы,  children, debt 

In [27]:
children = data_loan_users[['children', 'debt']]
print(children.head(10))

   children  debt
0         1     0
1         1     0
2         0     0
3         3     0
4         0     0
5         0     0
6         0     0
7         0     0
8         2     0
9         0     0


------------

# Комментарии

Результат этого дествия пригодится в дальнейшем, да?

------------

Определение зависимости между количеством детей и возвратом долга


Так как в таблице количество детей может быть разное, разделим заемщиков по группа:
    бездетные 0 детей
    с детьми 1-2 ребенка
    многодетные 3 и более детей

In [28]:
def class_children(children):
    if children == 0:
        return 'бездетные'
    if children >=1 and children < 3:
        return 'с детьми'
    return 'многодетные'
children['class_children'] = children['children'].apply(class_children)
print(children.head(10))

   children  debt class_children
0         1     0       с детьми
1         1     0       с детьми
2         0     0      бездетные
3         3     0    многодетные
4         0     0      бездетные
5         0     0      бездетные
6         0     0      бездетные
7         0     0      бездетные
8         2     0       с детьми
9         0     0      бездетные


/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


In [29]:
print(children['class_children'].value_counts())

бездетные      14091
с детьми        6907
многодетные      456
Name: class_children, dtype: int64


In [30]:
children_debt = children.groupby('class_children').agg({'debt': ['count', 'sum']})
print(children_debt)

                 debt      
                count   sum
class_children             
бездетные       14091  1063
многодетные       456    39
с детьми         6907   639


In [31]:
children_debt['share'] = children_debt['debt']['sum'] / children_debt['debt']['count']
print(children_debt.sort_values(by = 'share', ascending = False))

                 debt           share
                count   sum          
class_children                       
с детьми         6907   639  0.092515
многодетные       456    39  0.085526
бездетные       14091  1063  0.075438


Как видно из таблицы хуже всех отдают люди, у которых 1-2 ребенка, затем многодетные 
и лучше всех отдают семьи где нет детей. Однако стоит отметить что разница в доли между бездетными 
людьми с 1-2 ребенком составляет 1,7%


------------

# Комментарии

Хорошо, то есть наличие детей негативно влияет на вероятность возврата кредита в срок. Можешь это объяснить?

Немного больше комментариев к расчетам, пожалуста)

------------

Ответим на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?"

In [32]:
family_status_debt = data_loan_users[['family_status_id', 'family_status', 'debt']]
print(family_status_debt.head(10))

   family_status_id     family_status  debt
0                 0   женат / замужем     0
1                 0   женат / замужем     0
2                 0   женат / замужем     0
3                 0   женат / замужем     0
4                 1  гражданский брак     0
5                 1  гражданский брак     0
6                 0   женат / замужем     0
7                 0   женат / замужем     0
8                 1  гражданский брак     0
9                 0   женат / замужем     0


In [33]:
family_status_debt_1 = family_status_debt.groupby('family_status')['debt'].agg({'debt': ['count', 'sum']})
family_status_debt_1['share'] = family_status_debt_1['debt']['sum'] / family_status_debt_1['debt']['count']

print(family_status_debt_1.sort_values(by = 'share', ascending = False))

                        debt          share
                       count  sum          
family_status                              
Не женат / не замужем   2810  274  0.097509
гражданский брак        4151  388  0.093471
женат / замужем        12339  931  0.075452
в разводе               1195   85  0.071130
вдовец / вдова           959   63  0.065693


/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: using a dict on a Series for aggregation
is deprecated and will be removed in a future version. Use                 named aggregation instead.

    >>> grouper.agg(name_1=func_1, name_2=func_2)

  """Entry point for launching an IPython kernel.


Наиболее отвественно платят вдоцы и вдовы и наиболее плохо платят люди, у которых нет оформленных
отношений.

------------

# Комментарии

Хорошо) то есть отсутствие оформленных семейных отношений негативно влияет на вероятность возврата кредита в срок. Интересно, почему?

------------

Ответим на вопрос: :Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [34]:
print('Минимальное значение дохода', data_loan_users['total_income'].min())
print('Максимальное значение дохода', data_loan_users['total_income'].max())
print('Медианное значение дохода', data_loan_users['total_income'].median())
print('Среднее значение дохода', data_loan_users['total_income'].mean())

Минимальное значение дохода 20667
Максимальное значение дохода 2265604
Медианное значение дохода 142594.0
Среднее значение дохода 165319.57229421087


In [35]:
income_debt = data_loan_users[['income_type', 'total_income', 'debt']]
print(income_debt.head(10))

  income_type  total_income  debt
0   сотрудник        253875     0
1   сотрудник        112080     0
2   сотрудник        145885     0
3   сотрудник        267628     0
4   пенсионер        158616     0
5   компаньон        255763     0
6   компаньон        240525     0
7   сотрудник        135823     0
8   сотрудник         95856     0
9   сотрудник        144425     0


Выделем группы по общему доходу
- очень низкий доход до 30000
- низкий доход от 30 000 до 100 000
- средний доход от 100 000 до 500 000
- высокий доход от 500 000 до 1000 000
- очень высокий доход от 1000 000 и более

In [36]:
def class_total_income(income_type):
    if income_type <= 30000:
        return 'очень низкий доход'
    if income_type > 30000 and income_type < 100000:
        return 'низкий доход'
    if income_type > 100000 and income_type < 500000:
        return 'средний доход'
    if income_type > 500000 and income_type < 1000000:
        return 'высокий доход'
    return 'очень высокий доход'
income_debt['class_total_income'] = income_debt['total_income'].apply(class_total_income)
print(income_debt.head(10))

  income_type  total_income  debt class_total_income
0   сотрудник        253875     0      средний доход
1   сотрудник        112080     0      средний доход
2   сотрудник        145885     0      средний доход
3   сотрудник        267628     0      средний доход
4   пенсионер        158616     0      средний доход
5   компаньон        255763     0      средний доход
6   компаньон        240525     0      средний доход
7   сотрудник        135823     0      средний доход
8   сотрудник         95856     0       низкий доход
9   сотрудник        144425     0      средний доход


/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # This is added back by InteractiveShellApp.init_path()


In [37]:
class_total_income_debt = income_debt.groupby('class_total_income').agg({'debt': ['count', 'sum']})
print(class_total_income_debt)

                      debt      
                     count   sum
class_total_income              
высокий доход          197    12
низкий доход          4441   352
очень высокий доход     25     2
очень низкий доход      22     2
средний доход        16769  1373


In [38]:
class_total_income_debt['share'] = class_total_income_debt['debt']['sum'] / class_total_income_debt['debt']['count']
print(class_total_income_debt.sort_values(by = 'share', ascending = False))

                      debt           share
                     count   sum          
class_total_income                        
очень низкий доход      22     2  0.090909
средний доход        16769  1373  0.081877
очень высокий доход     25     2  0.080000
низкий доход          4441   352  0.079261
высокий доход          197    12  0.060914


Лучше всего платят люди с высоким и низким доходом, хуже всего возвращают долг люди с очень низким
и средним доходом

------------

# Комментарии

Хорошо, то есть зависимость есть) есть идеи, почему именно так?

Не забывай про комментарии)

------------

Дополнительно проанализируем зависимость долга от способа получения дохода и оценим в какую катеогрию попадают данные люди

In [39]:
income_type_income_debt = income_debt.groupby('income_type').agg({'total_income': ['min', 'max'], 'debt': ['count', 'sum']})
print(income_type_income_debt)
#print(income_type_income_deb)
print()
#print(income_type_dept_class_2)



                total_income            debt      
                         min      max  count   sum
income_type                                       
безработный            59956   202722      2     1
в декрете              53829    53829      1     1
госслужащий            29200   910451   1457    86
компаньон              28702  2265604   5078   376
пенсионер              20667   735103   3829   216
предприниматель       499163   499163      2     0
сотрудник              21367  1726276  11084  1061
студент                98201    98201      1     0



In [40]:
income_type_income_debt['share'] = income_type_income_debt['debt']['sum'] / income_type_income_debt['debt']['count']
print(income_type_income_debt.sort_values(by = 'share', ascending = False))

                total_income            debt           share
                         min      max  count   sum          
income_type                                                 
в декрете              53829    53829      1     1  1.000000
безработный            59956   202722      2     1  0.500000
сотрудник              21367  1726276  11084  1061  0.095724
компаньон              28702  2265604   5078   376  0.074045
госслужащий            29200   910451   1457    86  0.059025
пенсионер              20667   735103   3829   216  0.056412
предприниматель       499163   499163      2     0  0.000000
студент                98201    98201      1     0  0.000000


Как видно из таблицы, что наиболее плохо платят люди находящиеся в декрете и безработные, лучше всего платят студенты 
и предприниматели, однако стоит заметить что для анализа выборка по этим людям крайне маленькая и не репрезентативная. 
С другой стороны данные по категориям сотрудники, компаньены, госслужащие и пенсионеры достаточно объемны и могут давать достаточно точную информацию. Поэтому можно сказать, что худшими плательщиками по кредиту являются сотридники и компаньены, а лучшими пенсионеры и госслужащие. 

------------

# Комментарии

Отличная идея посмотреть на зависимость в таком разрезе)

Да, из-за того, что в некоторых категориях слишком мало наблюдений, выводы могут быть не корректны) 

------------

Как разные цели кредита влияют на его возврат в срок?

In [41]:
purpose_debt = data_loan_users[['purpose', 'debt']]
print(purpose_debt.head(10))

                  purpose  debt
0       операции с жильем     0
1  операции с автомобилем     0
2       операции с жильем     0
3   получение образования     0
4      проведение свадьбы     0
5       операции с жильем     0
6       операции с жильем     0
7   получение образования     0
8      проведение свадьбы     0
9       операции с жильем     0


In [42]:
class_purpose_debt = purpose_debt.groupby('purpose').agg({'debt': ['count', 'sum']})
print(class_purpose_debt)

                          debt     
                         count  sum
purpose                            
операции с автомобилем    4306  403
операции с жильем         5039  345
операции с недвижимостью  5772  437
получение образования     4013  370
проведение свадьбы        2324  186


In [43]:
class_purpose_debt['share'] = class_purpose_debt['debt']['sum'] / class_purpose_debt['debt']['count']
print(class_purpose_debt.sort_values(by = 'share', ascending = False))

                          debt          share
                         count  sum          
purpose                                      
операции с автомобилем    4306  403  0.093590
получение образования     4013  370  0.092200
проведение свадьбы        2324  186  0.080034
операции с недвижимостью  5772  437  0.075710
операции с жильем         5039  345  0.068466


Ноиболее ответсвенными плательщиками являются люди которые осуществляют операции с жильем и недвижемостью. Наихудшими плательщиками являются люди которые покупают автомобили и оплачивают получение образования.

------------

# Комментарии

Все верно) как бы ты объяснил эти выводы?

------------

Таким образом лучше всего двать кредит госслужащем или пенсионерам, у которых либо нет детей либо они являются многодетными, с годовым доходом: 
    1. От 30 000 до 100 000
    2. От 500 000 и более 
Наилучшими целями для возврата кредита являются операции с жильем или недвижемостью. Данные клиенты либо должны побывать в браке (разведены или овдоветь) либо находится в браке. 

Крайне осоторожно необходимо подходить к выдаче кредита людям, которые не оформляют свои отношения и имеют 1-2 ребенка, на цели преобретения автомобиля или получения образования с доходом либо очень низким или средним. 

------------

# Комментарии

Почему ты решил объединить группы людей по типу занятости, количеству детей и годовому доходу? Ты ведь проверял наличие зависимости в каждой категории отдельно, а не в разрезе нескольких факторов. И 2 пункта ниже содержат тот же подход.

------------

------------

# Комментарии

В общем ты молодец, выполнил все поставленные перед тобой задачи - это здорово)

Было бы лучше, если бы ты добавил общее описание проделанной работы к основным выводам.

Я бы попросила тебя давать больше комментариев по ходу выполнения работы (особенно к расчетам). Не нужно умалчивать о каких-то действиях и не описывать их. Мне важно видить как ты рассуждаешь, а не просто код. Почему ты делаешь так, а не иначе. На основе твоих рассуждений, я смогу давть тебе более полезные комментарии)

Посмотри про оформление [markdown ячеек](https://github.com/sandino/Markdown-Cheatsheet) . С помощью них можно сделать тетрадку более понятой - выделять основные шаги, изменять формат текста и т.д.

Поработай над комментариями внутри работы.

___Удачи!!!___

------------